src/tactic/positivity.lean

```diff
@@ -0,0 +1,350 @@
+  end
+
+/-! ### Core logic of the `positivity` tactic -/
+
+/-- Second base case of the `positivity` tactic.  Prove an expression `e` is positive/nonnegative by
+finding a hypothesis of the form `a < e` or `a ≤ e` in which `a` can be proved positive/nonnegative
+by `norm_num`. -/
+meta def compare_hyp (e p₂ : expr) : tactic strictness := do
+  p_typ ← infer_type p₂,
+  (lo, hi, strict₂) ← match p_typ with -- TODO also handle equality hypotheses
+  | `(%%lo ≤ %%hi) := pure (lo, hi, ff)
+  | `(%%hi ≥ %%lo) := pure (lo, hi, ff)
+  | `(%%lo < %%hi) := pure (lo, hi, tt)
+  | `(%%hi > %%lo) := pure (lo, hi, tt)
+  | _ := failed
+  end,
+  is_def_eq e hi,
+  strictness₁ ← norm_num.positivity lo,
+  match strictness₁, strict₂ with
+  | (positive p₁), tt := positive <$> mk_app ``lt_trans [p₁, p₂]
+  | (positive p₁), ff := positive <$> mk_app `lt_of_lt_of_le [p₁, p₂]
+  | (nonnegative p₁), tt := positive <$> mk_app `lt_of_le_of_lt [p₁, p₂]
+  | (nonnegative p₁), ff := nonnegative <$> mk_app `le_trans [p₁, p₂]
+  end
+
+/-- Attribute allowing a user to tag a tactic as an extension for `tactic.positivity`.  The main
+(recursive) step of this tactic is to try successively all the extensions tagged with this attribute
+on the expression at hand, and also to try the two "base case" tactics `tactic.norm_num.positivity`,
+`tactic.positivity.compare_hyp` on the expression at hand. -/
+@[user_attribute]
+meta def attr : user_attribute (expr → tactic strictness) unit :=
+{ name      := `positivity,
+  descr     := "extensions handling particular operations for the `positivity` tactic",
+  cache_cfg :=
+  { mk_cache := λ ns, do
+    { t ← ns.mfoldl
+        (λ (t : expr → tactic strictness) n, do
+          t' ← eval_expr (expr → tactic strictness) (expr.const n []),
+          pure (λ e, orelse' (t' e) (t e)))
+        (λ _, failed),
+      pure $ λ e, orelse'
+        (t e) $ orelse' -- run all the extensions on `e`
+          (norm_num.positivity e) $ -- directly try `norm_num` on `e`
+          -- loop over hypotheses and try to compare with `e`
+          local_context >>= list.foldl (λ tac h, orelse' tac (compare_hyp e h)) failed  },
+    dependencies := [] } }
+
+/-- Look for a proof of positivity/nonnegativity of an expression `e`; if found, return the proof
+together with a `strictness` stating whether the proof found was for strict positivity
+(`positive p`) or only for nonnegativity (`nonnegative p`). -/
+meta def core (e : expr) : tactic strictness := do
+  f ← attr.get_cache,
+  f e <|> fail "failed to prove positivity/nonnegativity"
+
+end positivity
+
+open positivity
+
+namespace interactive
+
+setup_tactic_parser
+
+/-- Tactic solving goals of the form `0 ≤ x` and `0 < x`.  The tactic works recursively according to
+the syntax of the expression `x`, if the atoms composing the expression all have numeric lower
+bounds which can be proved positive/nonnegative by `norm_num`.  This tactic either closes the goal
+or fails. -/
```

Comment:
Could you add a few examples of the tactic being used to the doc string? Copying from the tests is fine!